struct.js

```diff
@@ -0,0 +1,312 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+/* eslint max-len:[0, 120] */
+/* eslint max-statements:[0, 99] */
+'use strict';
+
+var assert = require('assert');
+var bufrw = require('bufrw');
+var TYPE = require('./TYPE');
+var UnexpectedFieldValueTypeidError = require('./errors').UnexpectedFieldValueTypeidError;
+var FieldRequiredError = require('./errors').FieldRequiredError;
+var skipField = require('./skip').skipField;
+
+var LengthResult = bufrw.LengthResult;
+var WriteResult = bufrw.WriteResult;
+var ReadResult = bufrw.ReadResult;
+
+function FieldSpec(def) {
+    var self = this;
+    self.id = def.id;
+    self.name = def.name;
+    self.required = def.required;
+    self.optional = def.optional;
+    self.annotations = def.annotations;
+    self.valueDefinition = def.valueType;
+    self.valueType = null;
+    self.constructDefaultValue = null; // def.defaultValue
+}
+
+FieldSpec.prototype.link = function link(spec) {
```

Comment:
In order to support forward references (references to types that were not mentioned before this type, and incidentally the containing type itself), this implementation works in two passes. The `compile()` pass constructs instances of all the types and their fields, and the `link()` pass connects the fields to the corresponding spec instances.